configuration/configurator/BuildSet.cpp

```diff
@@ -0,0 +1,246 @@
+{
+    int nLength = this->getBuildSetComponentCount();
+
+    for (int idx = 0; idx < nLength; idx++)
+    {
+        buildSetArray.append(this->getBuildSetComponentTypeName(idx));
+    }
+}
+
+void CBuildSetManager::getBuildSetServices(StringArray& buildSetArray) const
+{
+    int nLength = this->getBuildSetServiceCount();
+
+    for (int idx = 0; idx < nLength; idx++)
+    {
+        buildSetArray.append(this->getBuildSetServiceName(idx));
+    }
+}
+
+const char* CBuildSetManager::getBuildSetServiceName(int index) const
+{
+    return this->getBuildSetService(index)->getName();
+}
+
+const char* CBuildSetManager::getBuildSetServiceFileName(int index) const
+{
+    return this->getBuildSetService(index)->getSchema();
+}
+
+const char* CBuildSetManager::getBuildSetComponentTypeName(int index) const
+{
+    return this->getBuildSetComponent(index)->getName();
+}
+
+const char* CBuildSetManager::getBuildSetComponentFileName(int index) const
+{
+    return this->getBuildSetComponent(index)->getSchema();
+}
+
+const char* CBuildSetManager::getBuildSetProcessName(int index) const
+{
+    return this->getBuildSetComponent(index)->getProcessName();
+}
+
+bool CBuildSetManager::populateBuildSet()
+{
+    StringBuffer xpath;
+
+    if (m_buildSetTree.get() != NULL)
+        return false;
+
+    try
+    {
+        m_buildSetTree.set(createPTreeFromXMLFile(m_buildSetPath.str()));
+    }
+    catch(...)
+    {
+        return false;
+    }
+
+    xpath.appendf("./%s/%s/%s", XML_TAG_PROGRAMS, XML_TAG_BUILD, XML_TAG_BUILDSET);
+
+    Owned<IPropertyTreeIterator> iter = m_buildSetTree->getElements(xpath.str());
+
+    ForEach(*iter)
+    {
+        IPropertyTree* pTree = &iter->query();
+
+        if ( pTree->queryProp(XML_ATTR_PROCESS_NAME) == NULL || pTree->queryProp(XML_ATTR_OVERIDE) != NULL || ( (pTree->queryProp(XML_ATTR_DEPLOYABLE) != NULL && \
+                stricmp(pTree->queryProp(XML_ATTR_DEPLOYABLE), "no") == 0 && stricmp(pTree->queryProp(XML_ATTR_PROCESS_NAME), XML_TAG_ESPSERVICE) != 0) ) )
+            continue;
+
+        Owned<CBuildSet> pBuildSet = new CBuildSet(pTree->queryProp(XML_ATTR_INSTALLSET), pTree->queryProp(XML_ATTR_NAME), pTree->queryProp(XML_ATTR_PROCESS_NAME),\
+                                                   pTree->queryProp(XML_ATTR_SCHEMA), pTree->queryProp(XML_ATTR_DEPLOYABLE), pTree->queryProp(XML_ATTR_OVERIDE));
+
+        m_buildSetArray.append(*pBuildSet.getLink());
+    }
+    return true;
+}
+
+void CBuildSetManager::setBuildSetArray(const StringArray &strArray)
+{
+    m_buildSetArray.kill();
+
+    for (int idx = 0; idx < strArray.length(); idx++)
+    {
+        Owned<CBuildSet> pBSet = new CBuildSet(NULL, strArray.item(idx), NULL, strArray.item(idx));
+        assert (pBSet != NULL);
+        m_buildSetArray.append(*pBSet.getClear());
+    }
+}
+
+const char* CBuildSetManager::getBuildSetSchema(int index) const
+{
+    assert(index < m_buildSetArray.length());
```

Comment:
It's ok to keep in for debugging.  Should be excluded for release code through compile flag.